server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java

```diff
@@ -96,15 +100,13 @@ public Void visitFunction(Function function, Context context) {
                     if (usedRelationsFromBothEqOperatorArgs.size() < 2) {
                         context.isHashJoinPossible = false;
                     }
+                    context.insideEqOperator = saveInsideEqOperator;
                     break;
                 default:
                     if (context.insideEqOperator) {
                         for (Symbol arg : function.arguments()) {
                             arg.accept(this, context);
                         }
-                    } else {
-                        context.isHashJoinPossible = false;
-                        return null;
```

Comment:
Setting `context.isHashJoinPossible` to `true` means an equip-join condition has been observed, there could not be a situation that would revert this decision to my understanding.

Reply:
what about `a = b AND a + 1 > b` ?

Reply:
Thanks for reviewing. It seems like such join condition is a valid hash join. On current master it does become a hash join. (`HashJoinConditionSymbolsExtractor` chooses the first equality as the hash symbol.)

```
cr> explain verbose select * from t3 join t2 on t3.x = t2.y + 4 and t3.x = t2.y;                                                                                                              
+-----------------------------+----------------------------------------------------------+
| STEP                        | QUERY PLAN                                               |
+-----------------------------+----------------------------------------------------------+
| Initial logical plan        | Join[INNER | ((x = (y + 4)) AND (x = y))] (rows=unknown) |
|                             |   ├ Collect[doc.t3 | [x] | true] (rows=unknown)          |
|                             |   └ Collect[doc.t2 | [y, b] | true] (rows=unknown)       |
| optimizer_rewrite_join_plan | HashJoin[((x = (y + 4)) AND (x = y))] (rows=unknown)     |
|                             |   ├ Collect[doc.t3 | [x] | true] (rows=unknown)          |
|                             |   └ Collect[doc.t2 | [y, b] | true] (rows=unknown)       |
| Final logical plan          | HashJoin[((x = (y + 4)) AND (x = y))] (rows=unknown)     |
|                             |   ├ Collect[doc.t3 | [x] | true] (rows=unknown)          |
|                             |   └ Collect[doc.t2 | [y, b] | true] (rows=unknown)       |
+-----------------------------+----------------------------------------------------------+
EXPLAIN 3 rows in set (0.003 sec)
```

Also looking at the class level javadocs for `EquiJoinDetector` and `HashJoinConditionSymbolsExtractor` I get the impression that we are very lenient in allowing hash joins.

Reply:
I will open a new PR for this sorry for confusion.